### DISCUSSION TITLE
NRE throw during the startup when using latest Nightly build

### DISCUSSION THREAD

**Discussion by laolarou726** [AvaloniaUI/Avalonia#11491] [General]
**Describe the bug**

A NRE exception throw during the startup when using latest Avalonia Nightly Build:

```

 	Avalonia.Base.dll!Avalonia.Threading.DispatcherOperation.InvokeCore()
 	Avalonia.Base.dll!Avalonia.Threading.DispatcherOperation.Execute()
 	Avalonia.Base.dll!Avalonia.Threading.Dispatcher.ExecuteJob(Avalonia.Threading.DispatcherOperation job)
 	Avalonia.Base.dll!Avalonia.Threading.Dispatcher.ExecuteJobsCore()
 	Avalonia.Base.dll!Avalonia.Threading.Dispatcher.Signaled()
 	Avalonia.Win32.dll!Avalonia.Win32.Win32DispatcherImpl.DispatchWorkItem()
 	Avalonia.Win32.dll!Avalonia.Win32.Win32Platform.WndProc(nint hWnd, uint msg, nint wParam, nint lParam)
 	[本机到托管的转换]	
 	[托管到本机的转换]	
 	Avalonia.Win32.dll!Avalonia.Win32.Win32DispatcherImpl.RunLoop(System.Threading.CancellationToken cancellationToken)
 	Avalonia.Base.dll!Avalonia.Threading.DispatcherFrame.Run(Avalonia.Threading.IControlledDispatcherImpl impl)
 	Avalonia.Base.dll!Avalonia.Threading.Dispatcher.PushFrame(Avalonia.Threading.DispatcherFrame frame)
 	Avalonia.Base.dll!Avalonia.Threading.Dispatcher.MainLoop(System.Threading.CancellationToken cancellationToken)
 	Avalonia.Controls.dll!Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(string[] args)
 	Avalonia.Controls.dll!Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(Avalonia.AppBuilder builder, string[] args, Avalonia.Controls.ShutdownMode shutdownMode)
>	LauncherX.Avalonia.dll!LauncherX.Avalonia.Program.Main(string[] args) Line 14	C#

```

**Expected behavior**

The main window should show.

**Screenshots**

<img width="958" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/25716486/e74fdea7-70e4-4c8a-87e3-3881bb5b23ff">

**Desktop (please complete the following information):**

- OS: Windows 11 22H2
- Version: 11.0.999-cibuild0035015-beta

**Comment by laolarou726** [comment]
@maxkatz6 It's really hard for me to create a repro because I have no idea what is going on here. I just updated to the latest nightly and then things are falling apart.

**Comment by laolarou726** [comment]
It seems the problem is because the code doesn't ensure `_currentItemList` always has value.

<img width="1010" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/25716486/aec318b9-d57e-4fff-b90c-09b3ecd79314">

**Comment by laolarou726** [comment]
@kekekeks It seems the bug is coming from this commit: [Composition aware resources](https://github.com/AvaloniaUI/Avalonia/pull/11340). Where file `RenderDataDrawingContext.cs` at [Line 83](https://github.com/AvaloniaUI/Avalonia/blob/6c09a0f46b2dda8e1274081cba3866a1bf06adf9/src/Avalonia.Base/Rendering/Composition/Drawing/RenderDataDrawingContext.cs#L83) does not ensure the list always has the value. Could you please take a look at it?

**Comment by timunie** [comment]
@laolarou726 it really would be best if you find a way to reproduce it. Hard for us to find a solution otherwise. Maybe comment out line by line in your App until the startup fails. That should bring us a step further.

**Comment by laolarou726** [comment]
@timunie @kekekeks  I disabled "only my code" and then I got some more information about this issue:

<img width="882" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/25716486/4be0f0c5-3a4d-4612-b8c4-2f85f235a71a">
<img width="779" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/25716486/3ab207a2-9cdc-40c3-9819-f209ab8fc150">

It seems this exception is from this line of XAML:

<img width="715" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/25716486/90084d59-1f50-438b-9994-0993764161d9">

But I didn't see anything strange here.

**Comment by timunie** [comment]
Probably the binding to the Foreground. The control could be null during init.

**Comment by laolarou726** [comment]
> Probably the binding to the Foreground. The control could be null during init.

@kekekeks @maxkatz6 @timunie 

Found it, by removing all the binding from this control can solve this issue, is this intended or a bug? This never happens in previous releases. Also, controls that don't set the foreground explicitly will also break the app:

<img width="1089" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/25716486/47ab0fb6-7469-4d9c-942f-afd7482ac8bf">

**Comment by maxkatz6** [comment]
What control is it? I can't find one with IconPack control. Only wpf one.

**Comment by maxkatz6** [comment]
> Also, controls that don't set the foreground explicitly will also break the app:

I don't it's true, as there are plenty of controls without foreground being set but working. Bug again, can't tell without knowing for sure what's wrong.

**Comment by laolarou726** [comment]
> What control is it? I can't find one with IconPack control. Only wpf one.

I implemented this control by myself.

**Comment by maxkatz6** [comment]
Does it have any rendering code inside? I.e. Render method override or something like this.

**Comment by laolarou726** [comment]
> Does it have any rendering code inside? I.e. Render method override or something like this.

Nope, it only has a StyledProperty and a style resource.

**Comment by timunie** [comment]
@laolarou726 please provide a very minimal sample to reproduce. This is the best chance you have to get the issue closed. Without being able to reproduce, how should we judge if something is wrong?!

**Comment by timunie** [comment]
@laolarou726 let me convert this into a discussion for now. If we find the actual bug, a new issue can be opened, okay?

**Comment by laolarou726** [comment]
@timunie @kekekeks @maxkatz6  I just successfully created a repro for this issue! Here is the link to the repro: 

https://github.com/laolarou726/NullBindingTest

Sorry for the delay again, but I really need some time to figure out the issue here, hope you can understand that, by adding more controls to the view, the exception will throw.

**Comment by laolarou726** [comment]
@timunie Could you please convert it back to an issue please? Thank you so much!

**Comment by llfab** [comment]
I created a minimal sample showing the failure: https://github.com/llfab/Samples/tree/main/AvaloniaAnimationTest

In detail: 
```csharp
    public class MyCustomControl : Control
    {
        private readonly Stack<DrawingContext.PushedState> _pushedStates = new Stack<DrawingContext.PushedState>();

        public override void Render(DrawingContext context)
        {
            context.DrawRectangle(new SolidColorBrush(Colors.Green), null, new Rect(Bounds.Size));

            using (context.PushClip(new Rect(Bounds.Size)))
            {
                _pushedStates.Push(context.PushTransform(Matrix.Identity));
                _pushedStates.Push(context.PushClip(new Rect(Bounds.Size)));
                _pushedStates.Push(context.PushOpacity(1, new Rect(Bounds.Size)));

                // If you uncommment below drawing code, everything works
                //SolidColorBrush brush = new SolidColorBrush(Colors.Red);
                //IPen pen = new Pen(brush, 2);
                //context.DrawLine(pen, new Point(0, 0), new Point(300, 300));

                _pushedStates.Pop().Dispose();
                _pushedStates.Pop().Dispose(); // < this here fails
                _pushedStates.Pop().Dispose();
            }
        }
    }
```
The Dispose() of the popped state fails if you don't do any drawing after the last push. If you uncomment the lines and draw stuff, then everythings works fine.

If you leave it commented _currentItemsList of the next item is null after the first pop.

**Reply by timunie** [reply]
How did you find out this? I think this is really useful sample. Thank you so much ❤️

**Reply by llfab** [reply]
Well, I simply followed the path of our nice LayeredCanvas control which does all these things and now fails. The problem was mainly to just filter out the pure Avalonia DrawingContext calls out of our way more complex stack and compare our custom control debug state with the sample debug state...
Yes, it took about 60 min to get the sample into that state, because who would believe that may have to do with "not drawing things"...

**Comment by llfab** [comment]
Without drawing code...

![image](https://github.com/AvaloniaUI/Avalonia/assets/31031996/57e0c0a0-3f44-4501-8756-09f5643e2dc5)

and with drawing code...

![image](https://github.com/AvaloniaUI/Avalonia/assets/31031996/8f76d1fc-aac3-4056-9187-b68272b4c3c8)
